pkg/optin/optin.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2021 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package optin
+
+import (
+	"regexp"
+	"sort"
+	"strings"
+
+	generator "k8s.io/kube-state-metrics/v2/pkg/metric_generator"
+)
+
+// MetricFamilyFilter filters metric families which are defined as opt-in by their generator.FamilyGenerator
+type MetricFamilyFilter struct {
+	metrics []*regexp.Regexp
+}
+
+// Test tests if a given generator is an opt-in metric family and was passed as an opt-in metric family at startup
+func (filter MetricFamilyFilter) Test(generator generator.FamilyGenerator) bool {
+	if generator.OptIn {
+		for _, metric := range filter.metrics {
+			if metric.MatchString(generator.Name) {
+				return true
+			}
+		}
+		return false
+	}
+	return true
+}
+
+// Status returns the metrics contained within the filter as a comma-separated string
+func (filter MetricFamilyFilter) Status() string {
+	asStrings := make([]string, 0)
+	for _, metric := range filter.metrics {
+		asStrings = append(asStrings, metric.String())
+	}
+	// sort the strings for the sake of testing so that the resulting status is consistent
```

Comment:
Should the sort be done in the test?

Reply:
In the test, `(optin.MetricFamilyFilter).metrics` is passed in the order in which I expect the resulting status to be (and check on), but the resulting status does not always (sporadically) result in the same order as I passed in the metrics. Thus I added the sort in the method itself for the sake of testing but also to keep a consistent result in the logs.

If you really feel like this is not in place here I could try the following.
```go
asStrings := make([]string, len(filter.metrics))
for i, metric := range filter.metrics {
    asStrings[i] = metric.String()
}
return strings.Join(asStrings, ", ")
```

One downside of this (in my opinion) is that rather than being explicit we use a really specific way of creating the `asStrings` slice without explaining why or making it explicit using e.g. `sort.Strings`, resulting in the possibility that this is changed in the future and resulting in a test "randomly" failing.

Reply:
I see, if it was only the test that depended on the sorted order, then the test should be doing the sorting. But in this case it makes sense to keep the sort here for consistency in the logs. I would just adjust the comment to reflect that we are sorting to provide a consistent ux instead of making a test pass.

Reply:
I will change the comment to reflect this. Thanks for the feedback! 👍🏼